Implement CustomExecution extension API

shyshywhy opened the pull request:
Fixes: https://github.com/eclipse-theia/theia/issues/7185
Fixes: https://github.com/eclipse-theia/theia/issues/8767

#### What it does
Implement CustomExecution extension API, and fixed issue: https://github.com/eclipse-theia/theia/issues/7185, https://github.com/eclipse-theia/theia/issues/8767

#### How to test
Use Gradle Tasks extension to test.

#### Review checklist

- [x] as an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

shyshywhy commented:
![image](https://user-images.githubusercontent.com/18496567/110905076-e05ae480-8344-11eb-8f4f-61b8b75be123.png)

azatsarynnyy commented:
Hello @shyshywhy,
thank you for your contribution to Theia project!

`eclipsefdn/eca` status check isn't passed which means you have to sign an Eclipse Contributor Agreement before contributing the code.
By the following link https://wiki.eclipse.org/Development_Resources/Contributing_via_Git#Eclipse_Contributor_Agreement
you can find all the required information on how to sign an ECA.

shyshywhy commented:
@azatsarynnyy Ok, I fixed it.

vince-fugnitto commented on packages/task/src/node/custom/custom-task-runner-backend-module.ts:
```diff
@@ -0,0 +1,37 @@
+/********************************************************************************
+ * Copyright (C) 2018 Red Hat, Inc. and others.
```
@shyshywhy are you from `redhat`? We should update the file (if new) to your company and current year.

vince-fugnitto commented on packages/task/src/node/custom/custom-task-runner.ts:
```diff
@@ -0,0 +1,77 @@
+/********************************************************************************
+ * Copyright (C) 2017-2019 Ericsson and others.
```
Similar to the above `redhat` comment.

vince-fugnitto commented on packages/terminal/src/browser/terminal-widget-impl.ts:
```diff
@@ -36,7 +36,7 @@ import { TerminalSearchWidgetFactory, TerminalSearchWidget } from './search/term
 import { TerminalThemeService } from './terminal-theme-service';
 import { CommandLineOptions, ShellCommandBuilder } from '@theia/process/lib/common/shell-command-builder';
-
+import { Key } from '@theia/core/lib/browser/keys';
```
minor: preserve the new line after the import and before the exported const.

kittaakos commented:
I will have time for a full-blown review and verification tomorrow. Please bear with me. Thanks!

kittaakos commented:
I am verifying the PR with the `task-provider-sample` from [here](https://github.com/microsoft/vscode-extension-samples/tree/main/task-provider-sample).

Steps:
 - `Run Task...`
 - `custombuildscript: 32`.
 - `Continue without scanning the task output`.
 - Task terminates, see the `Terminal will be reused by task`, press `Enter`.
```
logger-protocol.ts:112 root ERROR The ws channel does not exist 23
log @ logger-protocol.ts:111
(anonymous) @ logger-frontend-module.ts:40
(anonymous) @ logger.ts:311
(anonymous) @ logger.ts:303
Promise.then (async)
../../packages/core/lib/common/logger.js.Logger.log @ logger.ts:298
log @ logger.ts:44
../../packages/core/lib/common/messaging/abstract-connection-provider.js.AbstractConnectionProvider.handleIncomingRawMessage @ abstract-connection-provider.ts:112
WebSocketConnectionProvider.socket.onmessage @ ws-connection-provider.ts:63
ReconnectingWebSocket._handleMessage @ reconnecting-websocket-mjs.js:172
main.js:904 Uncaught (in promise) Error: Connection got disposed.
    at Object.dispose (main.js:904)
    at Object.disposable.dispose (disposable.ts:96)
    at DisposableCollection.../../packages/core/lib/common/disposable.js.DisposableCollection.dispose (disposable.ts:74)
    at DisposableCollection.disposable.dispose (disposable.ts:96)
    at DisposableCollection.disposable.dispose (disposable.ts:96)
    at DisposableCollection.../../packages/core/lib/common/disposable.js.DisposableCollection.dispose (disposable.ts:74)
    at TerminalWidgetImpl.../../packages/core/lib/browser/widgets/widget.js.BaseWidget.dispose (widget.ts:68)
    at TerminalWidgetImpl.push.../../packages/terminal/lib/browser/terminal-widget-impl.js.TerminalWidgetImpl.dispose (terminal-widget-impl.ts:603)
    at terminal-widget-impl.ts:695
    at HTMLTextAreaElement.<anonymous> (widget.ts:226)
```

shyshywhy commented:
> I am verifying the PR with the `task-provider-sample` from [here](https://github.com/microsoft/vscode-extension-samples/tree/main/task-provider-sample).
> 
> Steps:
> 
> * `Run Task...`
> * `custombuildscript: 32`.
> * `Continue without scanning the task output`.
> * Task terminates, see the `Terminal will be reused by task`, press `Enter`.
> 
> ```
> logger-protocol.ts:112 root ERROR The ws channel does not exist 23
> log @ logger-protocol.ts:111
> (anonymous) @ logger-frontend-module.ts:40
> (anonymous) @ logger.ts:311
> (anonymous) @ logger.ts:303
> Promise.then (async)
> ../../packages/core/lib/common/logger.js.Logger.log @ logger.ts:298
> log @ logger.ts:44
> ../../packages/core/lib/common/messaging/abstract-connection-provider.js.AbstractConnectionProvider.handleIncomingRawMessage @ abstract-connection-provider.ts:112
> WebSocketConnectionProvider.socket.onmessage @ ws-connection-provider.ts:63
> ReconnectingWebSocket._handleMessage @ reconnecting-websocket-mjs.js:172
> main.js:904 Uncaught (in promise) Error: Connection got disposed.
>     at Object.dispose (main.js:904)
>     at Object.disposable.dispose (disposable.ts:96)
>     at DisposableCollection.../../packages/core/lib/common/disposable.js.DisposableCollection.dispose (disposable.ts:74)
>     at DisposableCollection.disposable.dispose (disposable.ts:96)
>     at DisposableCollection.disposable.dispose (disposable.ts:96)
>     at DisposableCollection.../../packages/core/lib/common/disposable.js.DisposableCollection.dispose (disposable.ts:74)
>     at TerminalWidgetImpl.../../packages/core/lib/browser/widgets/widget.js.BaseWidget.dispose (widget.ts:68)
>     at TerminalWidgetImpl.push.../../packages/terminal/lib/browser/terminal-widget-impl.js.TerminalWidgetImpl.dispose (terminal-widget-impl.ts:603)
>     at terminal-widget-impl.ts:695
>     at HTMLTextAreaElement.<anonymous> (widget.ts:226)
> ```

I fixed it, and fixed some problems found by my own test.

kittaakos commented:
> I fixed it, and fixed some problems found by my own test.

Thank you so much, you work blazing fast. I cannot catch up with your pace... I added a few comments here and there, none of them are crucial. Can you please look into them?

I am going to fetch your changes and keep on with the verification. I noticed an issue with the missing `exitCode` and `signal`. Maybe you have already fixed them.

kittaakos commented on packages/task/src/node/custom/custom-task-runner.ts:
```diff
@@ -0,0 +1,77 @@
+                context: ctx,
+                config: tskConfig,
+                label: tskConfig.label,
+                process: terminal,
```
Or ignore it :)

kittaakos commented on packages/task/src/node/custom/custom-task.ts:
```diff
@@ -0,0 +1,73 @@
+            config: this.options.config,
+            terminalId: this.process.id,
+            processId: this.process.id,
+            code: exitCode,
```
I think this should be:
```ts
  code: exitCode || 0,
```

VS Code works a bit differently, nothing happens on the callback:
https://github.com/microsoft/vscode/blob/d1e5c8cddf4fba06aad2e4cffd5b146f2de10700/src/vs/workbench/contrib/tasks/browser/terminalTaskSystem.ts#L372-L375

and we want to avoid logging an error on custom task execution here:
https://github.com/eclipse-theia/theia/blob/be686eef3e394bc521480fe8203aa7467e2cb98b/packages/task/src/browser/task-service.ts#L313

What do you think?

shyshywhy replied:
I think so too!

shyshywhy commented on packages/task/src/node/custom/custom-task.ts:
```diff
@@ -0,0 +1,78 @@
+        };
+    }
+
+    public callbackTaskComplete(exitCode: number | undefined): MaybePromise<void> {
```
Sorry, in TypeScript, the members are all public by default, I write a lot of java and I am confused.😊

shyshywhy commented on packages/plugin-ext/src/plugin/tasks/tasks.ts:
```diff
@@ -59,7 +64,24 @@ export class TasksExtImpl implements TasksExt {
+                pty.onDidClose((e: number | void = undefined) => {
+                    // eslint-disable-next-line no-void
+                    this.proxy.$customExecutionComplete(execution.id, e === void 0 ? undefined : e);
+                });
```
yes, dispose should before of customExecutionComplete.

shyshywhy commented on packages/plugin-ext/src/plugin/types-impl.ts:
```diff
@@ -1630,12 +1630,35 @@ export class ShellExecution {
 }
 
+export class CustomExecution {
+    private _callback: (resolvedDefintion: theia.TaskDefinition) => Thenable<theia.Pseudoterminal>;
```
Yes, I refer to a lot of vscode codes.

shyshywhy commented:
@kittaakos I fixed the above problem and merged it into the previous commit

kittaakos commented:
> Fixes: #8767

It does not really fix this issue. Although `vscode.CustomExecution` works, the Gradle task view is not populated in Theia. It works in VS Code. I found these errors when testing the `vscode.CustomExecution` with the _Gradle Tasks_ extension:

```
plugin-contribution-handler.ts:382 Uncaught (in promise) Error: command 'gradle.refresh' not found
    at PluginContributionHandler.<anonymous> (plugin-contribution-handler.ts:382)
    at step (plugin-contribution-handler.ts:15)
    at Object.next (plugin-contribution-handler.ts:15)
    at plugin-contribution-handler.ts:15
    at new Promise (<anonymous>)
    at push.../../packages/plugin-ext/lib/main/browser/plugin-contribution-handler.js.__awaiter (plugin-contribution-handler.ts:15)
    at Object.execute (plugin-contribution-handler.ts:379)
    at CommandRegistry.<anonymous> (command.ts:294)
    at step (command.ts:15)
    at Object.next (command.ts:15)
plugin-contribution-handler.ts:382 Uncaught (in promise) Error: command 'gradle.runBuild' not found
    at PluginContributionHandler.<anonymous> (plugin-contribution-handler.ts:382)
    at step (plugin-contribution-handler.ts:15)
    at Object.next (plugin-contribution-handler.ts:15)
    at plugin-contribution-handler.ts:15
    at new Promise (<anonymous>)
    at push.../../packages/plugin-ext/lib/main/browser/plugin-contribution-handler.js.__awaiter (plugin-contribution-handler.ts:15)
    at Object.execute (plugin-contribution-handler.ts:379)
    at CommandRegistry.<anonymous> (command.ts:294)
    at step (command.ts:15)
    at Object.next (command.ts:15)
plugin-contribution-handler.ts:382 Uncaught (in promise) Error: command 'gradle.explorerFlat' not found
    at PluginContributionHandler.<anonymous> (plugin-contribution-handler.ts:382)
    at step (plugin-contribution-handler.ts:15)
    at Object.next (plugin-contribution-handler.ts:15)
    at plugin-contribution-handler.ts:15
    at new Promise (<anonymous>)
    at push.../../packages/plugin-ext/lib/main/browser/plugin-contribution-handler.js.__awaiter (plugin-contribution-handler.ts:15)
    at Object.execute (plugin-contribution-handler.ts:379)
    at CommandRegistry.<anonymous> (command.ts:294)
    at step (command.ts:15)
    at Object.next (command.ts:15)
```

```
root ERROR [hosted-plugin: 20438] Promise rejection not handled in one second: Error: argument does not match one of these constraints: arg instanceof constraint, arg.constructor === constraint, nor constraint(arg) === true , reason: Error: argument does not match one of these constraints: arg instanceof constraint, arg.constructor === constraint, nor constraint(arg) === true
root ERROR [hosted-plugin: 20438] With stack trace: Error: argument does not match one of these constraints: arg instanceof constraint, arg.constructor === constraint, nor constraint(arg) === true
    at validateConstraint (file:///Users/akos.kitta/git/theia/examples/electron/lib/vs/editor/editor.main.js:16287:19)
    at Object.validateConstraints (file:///Users/akos.kitta/git/theia/examples/electron/lib/vs/editor/editor.main.js:16262:13)
    at idOrCommand.handler (file:///Users/akos.kitta/git/theia/examples/electron/lib/vs/editor/editor.main.js:82213:29)
    at InstantiationService.invokeFunction (file:///Users/akos.kitta/git/theia/examples/electron/lib/vs/editor/editor.main.js:84046:27)
    at Object.execute (file:///Users/akos.kitta/git/theia/examples/electron/lib/68.bundle.js:601:64)
    at CommandRegistry.<anonymous> (file:///Users/akos.kitta/git/theia/examples/electron/lib/bundle.js:133591:62)
    at step (file:///Users/akos.kitta/git/theia/examples/electron/lib/bundle.js:133368:23)
    at Object.next (file:///Users/akos.kitta/git/theia/examples/electron/lib/bundle.js:133349:53)
    at fulfilled (file:///Users/akos.kitta/git/theia/examples/electron/lib/bundle.js:133340:58)
```

```
Uncaught Error: Language client is not ready yet

Error: Language client is not ready yet
    at T.sendRequest (/Users/akos.kitta/.theia/extensions/redhat.java-0.76.0/extension/dist/extension.js:1)
    at o (/Users/akos.kitta/.theia/extensions/redhat.java-0.76.0/extension/dist/extension.js:9)
    at Object.provideFoldingRanges (/Users/akos.kitta/.theia/extensions/redhat.java-0.76.0/extension/dist/extension.js:9)
    at FoldingProviderAdapter.provideFoldingRanges (/Users/akos.kitta/git/theia/packages/plugin-ext/lib/plugin/languages/folding.js:31)
    at /Users/akos.kitta/git/theia/packages/plugin-ext/lib/plugin/languages.js:427
    at LanguagesExtImpl.<anonymous> (/Users/akos.kitta/git/theia/packages/plugin-ext/lib/plugin/languages.js:156)
    at step (/Users/akos.kitta/git/theia/packages/plugin-ext/lib/plugin/languages.js:48)
    at Object.next (/Users/akos.kitta/git/theia/packages/plugin-ext/lib/plugin/languages.js:29)
    at /Users/akos.kitta/git/theia/packages/plugin-ext/lib/plugin/languages.js:23
    at new Promise (<anonymous>)
    at /Users/akos.kitta/git/theia/examples/electron/lib/vs/editor/editor.main.js:1607
bundle.js:133596 Uncaught (in promise) Error: The command '__plugin.view.title.action.gradle.explorerFlat' cannot be executed. There are no active handlers available for the command.
    at CommandRegistry.<anonymous> (bundle.js:133596)
    at step (bundle.js:133368)
    at Object.next (bundle.js:133349)
    at bundle.js:133343
    at new Promise (<anonymous>)
    at ../../packages/core/lib/common/command.js.__awaiter (bundle.js:133339)
    at CommandRegistry.../../packages/core/lib/common/command.js.CommandRegistry.executeCommand (bundle.js:133581)
    at TabBarToolbar._this.executeCommand (bundle.js:121921)
    at HTMLUnknownElement.callCallback (bundle.js:51456)
    at Object.invokeGuardedCallbackDev (bundle.js:51505)
```

kittaakos commented:
I found an error when playing with the _Gradle Tasks_ extension:
 - Clone https://github.com/jitpack/gradle-simple,
 - Open in a Theia workspace,
 - Build starts, Gradle config starts 👍,
 - See the thumbs-up in the status bar, click on it,
 - Nothing happens, see the error in the log.

```
Uncaught (in promise) Error: Run task config does not return after being started
    at TasksExtImpl.<anonymous> (/Users/akos.kitta/gi…/tasks/tasks.js:254)
    at step (/Users/akos.kitta/gi…n/tasks/tasks.js:48)
    at Object.next (/Users/akos.kitta/gi…n/tasks/tasks.js:29)
    at fulfilled (/Users/akos.kitta/gi…n/tasks/tasks.js:20)
    at processTicksAndRejections (/Users/akos.kitta/gi…s/task_queues.js:94)
```

kittaakos commented:
Unfortunately, there is another issue. When I install the _Language Support for Java(TM) by Red Hat_ VSX, I can check the build status task with the following command:

![Screen Shot 2021-03-23 at 13 23 03](https://user-images.githubusercontent.com/1405703/112145787-1a48a800-8bdb-11eb-9748-e3d865dcf834.jpg)

In VS Code, it shows the terminal that runs the task:
![Screen Shot 2021-03-23 at 13 24 48](https://user-images.githubusercontent.com/1405703/112145847-26cd0080-8bdb-11eb-92e4-b4f78ef42ebf.jpg)

In Theia, I always see this:
![Screen Shot 2021-03-23 at 13 25 09](https://user-images.githubusercontent.com/1405703/112145885-33515900-8bdb-11eb-960e-54e30f9e5a01.jpg)

But the task is not active anymore.

kittaakos commented:
The PR looks very promising, @shyshywhy. If you need assistance with the bug-fixes, let me know, and I try to chime in into the development. Thank you so much for your help so far 🙏

shyshywhy commented:
> > Fixes: #8767
> 
> It does not really fix this issue. Although `vscode.CustomExecution` works, the Gradle task view is not populated in Theia. It works in VS Code. I found these errors when testing the `vscode.CustomExecution` with the _Gradle Tasks_ extension:
> 
> ```
> plugin-contribution-handler.ts:382 Uncaught (in promise) Error: command 'gradle.refresh' not found
>     at PluginContributionHandler.<anonymous> (plugin-contribution-handler.ts:382)
>     at step (plugin-contribution-handler.ts:15)
>     at Object.next (plugin-contribution-handler.ts:15)
>     at plugin-contribution-handler.ts:15
>     at new Promise (<anonymous>)
>     at push.../../packages/plugin-ext/lib/main/browser/plugin-contribution-handler.js.__awaiter (plugin-contribution-handler.ts:15)
>     at Object.execute (plugin-contribution-handler.ts:379)
>     at CommandRegistry.<anonymous> (command.ts:294)
>     at step (command.ts:15)
>     at Object.next (command.ts:15)
> plugin-contribution-handler.ts:382 Uncaught (in promise) Error: command 'gradle.runBuild' not found
>     at PluginContributionHandler.<anonymous> (plugin-contribution-handler.ts:382)
>     at step (plugin-contribution-handler.ts:15)
>     at Object.next (plugin-contribution-handler.ts:15)
>     at plugin-contribution-handler.ts:15
>     at new Promise (<anonymous>)
>     at push.../../packages/plugin-ext/lib/main/browser/plugin-contribution-handler.js.__awaiter (plugin-contribution-handler.ts:15)
>     at Object.execute (plugin-contribution-handler.ts:379)
>     at CommandRegistry.<anonymous> (command.ts:294)
>     at step (command.ts:15)
>     at Object.next (command.ts:15)
> plugin-contribution-handler.ts:382 Uncaught (in promise) Error: command 'gradle.explorerFlat' not found
>     at PluginContributionHandler.<anonymous> (plugin-contribution-handler.ts:382)
>     at step (plugin-contribution-handler.ts:15)
>     at Object.next (plugin-contribution-handler.ts:15)
>     at plugin-contribution-handler.ts:15
>     at new Promise (<anonymous>)
>     at push.../../packages/plugin-ext/lib/main/browser/plugin-contribution-handler.js.__awaiter (plugin-contribution-handler.ts:15)
>     at Object.execute (plugin-contribution-handler.ts:379)
>     at CommandRegistry.<anonymous> (command.ts:294)
>     at step (command.ts:15)
>     at Object.next (command.ts:15)
> ```
> 
> ```
> root ERROR [hosted-plugin: 20438] Promise rejection not handled in one second: Error: argument does not match one of these constraints: arg instanceof constraint, arg.constructor === constraint, nor constraint(arg) === true , reason: Error: argument does not match one of these constraints: arg instanceof constraint, arg.constructor === constraint, nor constraint(arg) === true
> root ERROR [hosted-plugin: 20438] With stack trace: Error: argument does not match one of these constraints: arg instanceof constraint, arg.constructor === constraint, nor constraint(arg) === true
>     at validateConstraint (file:///Users/akos.kitta/git/theia/examples/electron/lib/vs/editor/editor.main.js:16287:19)
>     at Object.validateConstraints (file:///Users/akos.kitta/git/theia/examples/electron/lib/vs/editor/editor.main.js:16262:13)
>     at idOrCommand.handler (file:///Users/akos.kitta/git/theia/examples/electron/lib/vs/editor/editor.main.js:82213:29)
>     at InstantiationService.invokeFunction (file:///Users/akos.kitta/git/theia/examples/electron/lib/vs/editor/editor.main.js:84046:27)
>     at Object.execute (file:///Users/akos.kitta/git/theia/examples/electron/lib/68.bundle.js:601:64)
>     at CommandRegistry.<anonymous> (file:///Users/akos.kitta/git/theia/examples/electron/lib/bundle.js:133591:62)
>     at step (file:///Users/akos.kitta/git/theia/examples/electron/lib/bundle.js:133368:23)
>     at Object.next (file:///Users/akos.kitta/git/theia/examples/electron/lib/bundle.js:133349:53)
>     at fulfilled (file:///Users/akos.kitta/git/theia/examples/electron/lib/bundle.js:133340:58)
> ```
> 
> ```
> Uncaught Error: Language client is not ready yet
> 
> Error: Language client is not ready yet
>     at T.sendRequest (/Users/akos.kitta/.theia/extensions/redhat.java-0.76.0/extension/dist/extension.js:1)
>     at o (/Users/akos.kitta/.theia/extensions/redhat.java-0.76.0/extension/dist/extension.js:9)
>     at Object.provideFoldingRanges (/Users/akos.kitta/.theia/extensions/redhat.java-0.76.0/extension/dist/extension.js:9)
>     at FoldingProviderAdapter.provideFoldingRanges (/Users/akos.kitta/git/theia/packages/plugin-ext/lib/plugin/languages/folding.js:31)
>     at /Users/akos.kitta/git/theia/packages/plugin-ext/lib/plugin/languages.js:427
>     at LanguagesExtImpl.<anonymous> (/Users/akos.kitta/git/theia/packages/plugin-ext/lib/plugin/languages.js:156)
>     at step (/Users/akos.kitta/git/theia/packages/plugin-ext/lib/plugin/languages.js:48)
>     at Object.next (/Users/akos.kitta/git/theia/packages/plugin-ext/lib/plugin/languages.js:29)
>     at /Users/akos.kitta/git/theia/packages/plugin-ext/lib/plugin/languages.js:23
>     at new Promise (<anonymous>)
>     at /Users/akos.kitta/git/theia/examples/electron/lib/vs/editor/editor.main.js:1607
> bundle.js:133596 Uncaught (in promise) Error: The command '__plugin.view.title.action.gradle.explorerFlat' cannot be executed. There are no active handlers available for the command.
>     at CommandRegistry.<anonymous> (bundle.js:133596)
>     at step (bundle.js:133368)
>     at Object.next (bundle.js:133349)
>     at bundle.js:133343
>     at new Promise (<anonymous>)
>     at ../../packages/core/lib/common/command.js.__awaiter (bundle.js:133339)
>     at CommandRegistry.../../packages/core/lib/common/command.js.CommandRegistry.executeCommand (bundle.js:133581)
>     at TabBarToolbar._this.executeCommand (bundle.js:121921)
>     at HTMLUnknownElement.callCallback (bundle.js:51456)
>     at Object.invokeGuardedCallbackDev (bundle.js:51505)
> ```

For this issue, i have fixed, but not sure if the repair is appropriate.

In the Task, updateDefinitionBasedOnExecution will modify the type of TaskDefinition, resulting in plugin judgment errors.

```
private updateDefinitionBasedOnExecution(): void {
        if (this.taskExecution instanceof ProcessExecution) {
            Object.assign(this.taskDefinition, {
                type: 'process',
                id: this.taskExecution.computeId(),
                taskType: this.taskDefinition!.type
            });
        } else if (this.taskExecution instanceof ShellExecution) {
            Object.assign(this.taskDefinition, {
                type: 'shell',
                id: this.taskExecution.computeId(),
                taskType: this.taskDefinition!.type
            });
        } else if (this.taskExecution instanceof CustomExecution) {
            Object.assign(this.taskDefinition, {
                type: 'customExecution',
                id: this.taskExecution.computeId(),
                taskType: this.taskDefinition!.type
            });
        } else {
            Object.assign(this.taskDefinition, {
                type: '$empty',
                id: UUID.uuid4(),
                taskType: this.taskDefinition!.type
            });
        }
    }
``` 

so, i change it, but it’s not clear if it conflicts with theia design

shyshywhy commented:
> Unfortunately, there is another issue. When I install the _Language Support for Java(TM) by Red Hat_ VSX, I can check the build status task with the following command:
> 
> ![Screen Shot 2021-03-23 at 13 23 03](https://user-images.githubusercontent.com/1405703/112145787-1a48a800-8bdb-11eb-9748-e3d865dcf834.jpg)
> 
> In VS Code, it shows the terminal that runs the task:
> ![Screen Shot 2021-03-23 at 13 24 48](https://user-images.githubusercontent.com/1405703/112145847-26cd0080-8bdb-11eb-92e4-b4f78ef42ebf.jpg)
> 
> In Theia, I always see this:
> ![Screen Shot 2021-03-23 at 13 25 09](https://user-images.githubusercontent.com/1405703/112145885-33515900-8bdb-11eb-960e-54e30f9e5a01.jpg)
> 
> But the task is not active anymore.


I think can open another issue to track this issue.

The reason for this problem is that the repeated judgment of the task does not distinguish the client, and there is no recovery mechanism after the terminal is closed.

kittaakos commented:
> Although `vscode.CustomExecution` works, the Gradle task view is not populated in Theia.

👍 This works now:

![screencast 2021-03-30 10-04-12](https://user-images.githubusercontent.com/1405703/112955370-a3ae2c00-913f-11eb-9668-72b91a1af17a.gif)

kittaakos commented:
Thank you for the update, @shyshywhy. I will approve your changes and leave the PR open for a few more days if someone else wants to chime in and try it in action. I have not used the `vscode.CustomExecution` before, but I have tried the followings with your changeset:
 - https://github.com/microsoft/vscode-extension-samples/tree/main/task-provider-sample,
 - https://marketplace.visualstudio.com/items?itemName=richardwillis.vscode-gradle, and
 - https://marketplace.visualstudio.com/items?itemName=redhat.java.

Overall, it's a great improvement. Thank you so much for your help 👍 

----
There are a few issues I have noticed. We can fix them later:

 1. Somehow, I managed to get an error when opening the terminal for a task. I am not sure how to reproduce it:
```
command.ts:15 Uncaught (in promise) Error: Run task config does not return after being started
    at TasksExtImpl.<anonymous> (/Users/akos.kitta/git/theia/packages/plugin-ext/lib/plugin/tasks/tasks.js:254)
    at step (/Users/akos.kitta/git/theia/packages/plugin-ext/lib/plugin/tasks/tasks.js:48)
    at Object.next (/Users/akos.kitta/git/theia/packages/plugin-ext/lib/plugin/tasks/tasks.js:29)
    at fulfilled (/Users/akos.kitta/git/theia/packages/plugin-ext/lib/plugin/tasks/tasks.js:20)
    at processTicksAndRejections (internal/process/task_queues.js:94)
```

 2. When a task is finished, I can see the 👍 (thumbs-up icon) in the status bar, when I click on it, I see the notification with the action, but I cannot reopen the terminal for the task. (See [here](https://github.com/eclipse-theia/theia/issues/7185#issuecomment-633794410).)
![Screen Shot 2021-03-30 at 13 08 10](https://user-images.githubusercontent.com/1405703/112991775-cb18ef00-9167-11eb-94b9-d544941c13eb.jpg)

 3. If I am not mistaken, Theia does create an identifier for a task. VS Code has a [`KeyedTaskIdentifier`](https://github.com/microsoft/vscode/blob/142cbdbaad80700ce108984ae34028d52b8d7fe7/src/vs/workbench/contrib/tasks/common/tasks.ts#L447-L449) that can be used if two tasks are the "same". VS Code can report that a "same" task is already running, or here is the result of that task.

 4.  When trying the Gradle Tasks VS Code extension, I got this: (_unrelated to this PR_)
```
Error: Language client is not ready yet
    at T.sendRequest (/Users/akos.kitta/.t…dist/extension.js:1)
    at o (/Users/akos.kitta/.t…dist/extension.js:9)
    at Object.provideFoldingRanges (/Users/akos.kitta/.t…dist/extension.js:9)
    at FoldingProviderAdapter.provideFoldingRanges (/Users/akos.kitta/gi…uages/folding.js:31)
    at /Users/akos.kitta/gi…in/languages.js:427
    at LanguagesExtImpl.<anonymous> (/Users/akos.kitta/gi…in/languages.js:156)
    at step (/Users/akos.kitta/gi…gin/languages.js:48)
    at Object.next (/Users/akos.kitta/gi…gin/languages.js:29)
    at /Users/akos.kitta/gi…gin/languages.js:23
    at new Promise (<anonymous>)
    at errors.ts:26
```

 5. I got this once when I reloaded the browser example: (_I assume this is unrelated to `vscode.CustomExecution`_)
```
root ERROR Could not transfer 'README.md' file from 'redhat_java' Error: Request aborted
    at onaborted (/Users/akos.kitta/git/theia/node_modules/express/lib/response.js:1025:15)
    at Immediate.<anonymous> (/Users/akos.kitta/git/theia/node_modules/express/lib/response.js:1067:9)
    at processImmediate (internal/timers.js:439:21)
root ERROR Could not transfer 'README.md' file from 'richardwillis_vscode_gradle' Error: Request aborted
    at onaborted (/Users/akos.kitta/git/theia/node_modules/express/lib/response.js:1025:15)
    at Immediate.<anonymous> (/Users/akos.kitta/git/theia/node_modules/express/lib/response.js:1067:9)
    at processImmediate (internal/timers.js:439:21)
```

 6.  There was some issue with OVSX when I started the app: (_Seems to be unrelated to this PR_)
```
root ERROR [foo-bar.task-provider-samples]: failed to refresh, reason: Error: Extension with id foo-bar.task-provider-samples not found at https://open-vsx.org/api
    at VSXRegistryAPI.<anonymous> (http://localhost:3000/71.bundle.js:3165:31)
    at step (http://localhost:3000/71.bundle.js:3020:23)
    at Object.next (http://localhost:3000/71.bundle.js:3001:53)
    at fulfilled (http://localhost:3000/71.bundle.js:2992:58)
```

kittaakos reviewed (approved):
Thank you so much for your help, @shyshywhy 👍

vince-fugnitto reviewed (commented):
@shyshywhy do you mind rebasing the pull-request against the latest master? There are some changes which will be required due to the `re-export` pull-request: https://github.com/eclipse-theia/theia/pull/9124.

shyshywhy commented:
> @shyshywhy do you mind rebasing the pull-request against the latest master? There are some changes which will be required due to the `re-export` pull-request: #9124.

ok，its done.

RomanNikitenko commented on packages/task/src/node/task-server.ts:
```diff
@@ -89,7 +90,7 @@ export class TaskServerImpl implements TaskServer, Disposable {
 
     async run(taskConfiguration: TaskConfiguration, ctx?: string, option?: RunTaskOption): Promise<TaskInfo> {
-        const runner = this.runnerRegistry.getRunner(taskConfiguration.type);
+        const runner = this.runnerRegistry.getRunner(taskConfiguration.taskType);
```
@shyshywhy 
could you share the reason why `type` was replaced by `taskType`?

vince-fugnitto commented:
The commit seems to have broken a lot of functionality from `tasks`, should we revert for the time being?

- https://github.com/eclipse-theia/theia/blob/master/doc/pull-requests.md#reverting-pr

> If a PR causes regressions after landing then an author and maintainers have 2 days to resolve them after that a PR has to be reverted.

If the issues are not resolved before the release I think we should revert as to not release a broken task system.

tsmaeder commented:
@vince-fugnitto So what breakage is left?

vince-fugnitto commented:
> @vince-fugnitto So what breakage is left?

I see the following for the moment (possibly others):
- https://github.com/eclipse-theia/theia/issues/9365
- https://github.com/eclipse-theia/theia/issues/9366
- https://github.com/eclipse-theia/theia/pull/9207#pullrequestreview-637691473

RomanNikitenko commented:
https://github.com/eclipse-theia/theia/issues/9341

azatsarynnyy commented:
I fully agree with Vincent. It should be reverted according to our PR policy. As the PR originator hasn't responded for a week.

RomanNikitenko commented:
I'm not able to take any issue from the list above in progress soon, so I'm +1 for reverting...

azatsarynnyy commented:
> I fully agree with Vincent. It should be reverted according to our PR policy. As the PR originator hasn't responded for a week.

unless someone else is going to fix all the breakages before the release

azatsarynnyy commented:
On the other hand, we have almost two weeks before the release date...
So, I see two choices here:
- revert the PR
- or get the concrete assignees for all the issues found, who could help to fix all of them

tsmaeder commented:
I think we all agree this PR should not have been merged. The question now is what the way forward should be. Personally, I think we (probably me, since I already have a head start) should fix the issues, The PR was not huge and I think we could fix the issues in question within maybe 2 days. At least invest a couple of hours to see what the problem is before throwing out the feature.

vince-fugnitto commented:
> I think we all agree this PR should not have been merged. The question now is what the way forward should be. Personally, I think we (probably me, since I already have a head start) should fix the issues, The PR was not huge and I think we could fix the issues in question within maybe 2 days. At least invest a couple of hours to see what the problem is before throwing out the feature.

@tsmaeder since we have time before the release I think we can spend time to fix the regressions, and if ever we see that we run out of time, or there are too many regressions for the framework to be released we can think of reverting.

tsmaeder commented:
Sounds like a plan!

RomanNikitenko commented:
Great!
I can help with testing :-)

alvsan09 commented:
The PR #9189 merged 5 commits to support the Tasks customExecution API,
one of the commits ("Fixed task type is modified") is the one that has caused cascading issues related to task type,
The PR #9373, removes only that particular problematic commit and two other trying to work around the main issue.

The good part is that removing that particular commit does not remove the support for customExecution,
the impacted scenario is #8767 where gradle tasks are not visible.

if we consider PR #9373 as a good base, we can then concentrate on fixing #8767.
